linux/dm-sworndisk/src/bio.rs

```diff
@@ -0,0 +1,350 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2023 Ant Group CO., Ltd.
+
+//! Block I/O.
+
+use core::{
+    fmt,
+    ops::{Deref, DerefMut, Range},
+    ptr::NonNull,
+};
+use kernel::{prelude::*, types::Opaque};
+
+use super::block_device::{HostBlockDevice, BLOCK_SECTORS, BLOCK_SIZE};
+
+// See definition at `include/linux/gfp_types.h`.
+const GFP_NOIO: bindings::gfp_t = bindings::___GFP_DIRECT_RECLAIM | bindings::___GFP_KSWAPD_RECLAIM;
+
+/// A handle to kernel's struct `bio`.
+pub(crate) struct Bio {
+    inner: NonNull<bindings::bio>,
+}
+
+// SAFETY: we get a reference to kernel's `bio` by `bio_get`, so it won't disappear
+// and can be safely transferred across threads.
+unsafe impl Send for Bio {}
+
+impl Bio {
+    /// Constructs a `Bio` from raw pointer.
+    ///
+    /// # Safety
+    ///
+    /// User must provide a valid pointer to the kernel's `bio`.
+    pub unsafe fn from_raw(ptr: *mut bindings::bio) -> Self {
+        // Get a reference to `bio`, so it won't disappear. And `bio_put` is
+        // called when it's been dropped.
+        bindings::bio_get(ptr);
+
+        Self {
+            inner: NonNull::new_unchecked(ptr),
+        }
+    }
+
+    /// Allocates a `Bio` for the target `HostBlockDevice`.
+    ///
+    /// The bio operation is specified by `op`, targeting a given block region,
+    /// while its I/O buffer is configured by `BioVec`.
+    pub fn alloc(
+        bdev: &HostBlockDevice,
+        bio_op: BioOp,
+        region: Range<usize>,
+        io_buffer: &[BioVec],
+    ) -> Result<Self> {
+        // SAFETY: `bdev.as_ptr()` is valid pointer to `block_device`, `bio_alloc_bioset`
+        // returns a pointer to new `bio`, or `NULL` on failure.
+        let bio_ptr = unsafe {
+            bindings::bio_alloc_bioset(
+                bdev.as_ptr(),
+                io_buffer.len() as _,
+                bio_op as _,
+                GFP_NOIO,
+                &mut bindings::fs_bio_set,
+            )
+        };
+        if bio_ptr.is_null() {
+            return Err(ENOMEM);
+        }
+
+        let mut buf_len = 0usize;
+        for bio_vec in io_buffer {
+            // SAFETY: `bio_vec` contains valid pointer to `page`, `bio_ptr` is
+            // non-null and valid.
+            unsafe {
+                let page = bindings::virt_to_page(bio_vec.virt_addr as _);
+                bindings::__bio_add_page(bio_ptr, page, bio_vec.len as _, bio_vec.offset as _);
+                buf_len += bio_vec.len;
+            }
+        }
+
+        let region_bytes = region.len() * BLOCK_SIZE;
+        if region_bytes != buf_len {
```

Comment:
Can we put length mismatch check in the front of this function?

Reply:
Sure.